src/types/dict.rs

```diff
@@ -180,6 +180,12 @@ pub trait PyDictMethods<'py>: crate::sealed::Sealed {
     /// so long as the set of keys does not change.
     fn iter(&self) -> BoundDictIterator<'py>;
 
+    /// Iterates over the contents of this dictionary while preventing other threads from modifying it.
+    #[cfg(Py_GIL_DISABLED)]
+    fn locked_for_each<F>(&self, closure: F) -> PyResult<()>
+    where
+        F: Fn(Bound<'py, PyAny>, Bound<'py, PyAny>) -> PyResult<()>;
```

Comment:
I think this API description isn't quite correct; the critical section may be released by the GC and then other threads may take the opportunity to modify the dictionary before this thread can reacquire the critical section.

All the critical section really guarantees is that this thread can safely call `PyDict_Next` and take ownership of the borrowed references before any other critical section gets nested.

I guess it is potentially nice to expose this API to users, but it's a bit tricky to find a good wording.

Reply:
It's a bit stronger than that, but I agree that it's tricky to find the right wording. Threads won't pause just anywhere for the GC -- they pause at pretty much the same places where the GIL could have been released in the default build.

So in practice, this means that if `closure()` doesn't make any API calls that could have released the GIL internally, then the dictionary will remain locked for the entire iteration.

Reply:
I have updated the api description, let me know if this is what you were looking for.